mongo/job_store.go

```diff
@@ -0,0 +1,170 @@
+const jobsCol = "jobs"
+
+// locked jobs collection name
+const jobsLockCol = "jobs_locks"
+
+//JobStore is a type that contains an implementation of the MongoJobStorer interface, which can be used for creating and getting Job resources.
+//It also represents a simplistic MongoDB configuration, with session, health and lock clients
+type JobStore struct {
+	Session      *mgo.Session
+	URI          string
+	Database     string
+	Collection   string
+	client       *dpMongoHealth.Client
+	healthClient *dpMongoHealth.CheckMongoClient
+	lockClient   *dpMongoLock.Lock
+}
+
+func (m *JobStore) CreateJob(ctx context.Context, id string) (job models.Job, err error) {
+	log.Event(ctx, "creating job in mongo DB", log.Data{"id": id})
+
+	// If an empty id was passed in, return an error with a message.
+	if id == "" {
+		return models.Job{}, errors.New("id must not be an empty string")
+	}
+
+	//Create a Job that's populated with default values of all its attributes
+	newJob := models.NewJob(id)
+
+	s := m.Session.Copy()
+	defer s.Close()
+	var jobToFind models.Job
+
+	//Check that the jobs collection does not already contain the id as a key
+	err = s.DB(m.Database).C(jobsCol).Find(bson.M{"id": id}).One(&jobToFind)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			//this means we CAN insert the job as it does not already exist
+			err = s.DB(m.Database).C(m.Collection).Insert(newJob)
+			if err != nil {
+				return models.Job{}, errors.New("error inserting job into mongo DB")
+			}
+			log.Event(ctx, "adding job to jobs collection", log.Data{"Job details: ": newJob})
+		} else {
+			//an unexpected error has occurred
+			return models.Job{}, err
+		}
+	} else {
+		//no error means that it found a job already exists with the id we're trying to insert
+		return models.Job{}, errors.New("id must be unique")
+	}
+
+	return newJob, nil
+}
+
+// Init creates a new mgo.Session with a strong consistency and a write mode of "majority".
+func (m *JobStore) Init(ctx context.Context) (err error) {
+	if m.Session != nil {
+		return errors.New("session already exists")
+	}
+
+	// Create session
+	if m.Session, err = mgo.Dial(m.URI); err != nil {
+		return err
+	}
+	m.Session.EnsureSafe(&mgo.Safe{WMode: "majority"})
+	m.Session.SetMode(mgo.Strong, true)
+
+	databaseCollectionBuilder := make(map[dpMongoHealth.Database][]dpMongoHealth.Collection)
+	databaseCollectionBuilder[(dpMongoHealth.Database)(m.Database)] = []dpMongoHealth.Collection{(dpMongoHealth.Collection)(m.Collection), jobsLockCol}
+	// Create client and healthClient from session
+	m.client = dpMongoHealth.NewClientWithCollections(m.Session, databaseCollectionBuilder)
+	m.healthClient = &dpMongoHealth.CheckMongoClient{
+		Client:      *m.client,
+		Healthcheck: m.client.Healthcheck,
+	}
+
+	// Create MongoDB lock client, which also starts the purger loop
+	m.lockClient = dpMongoLock.New(ctx, m.Session, m.Database, jobsCol)
+	return nil
+}
+
+//AcquireJobLock tries to lock the provided jobID.
```

Comment:
Minor formatting suggestion, some comments have a space between the `//` and the text, I think it looks cleaner.

Reply:
done